File: Valid Palindrome - Leetcode 125/Valid Palindrome - Leetcode 125.js
// Solution 1:  Time: O(n), Space: O(n)
const isPalindrome = function (s) {
  const filter = s.match(/[a-zA-Z0-9]/g);

  if (!filter) return true;

  const str = filter.join("").toLowerCase();

  const reversedStr = filter.reverse().join("").toLowerCase();

  if (str === reversedStr) return true;

  return false;
};

// Solution 2: Time: O(n) , Space: O(1)
const isPalindrome = function (s) {
  let l = 0,
    r = s.length - 1;

  while (l < r) {
    if (!s[l].match(/^[a-z0-9]+$/i)) {
      l++;
      continue;
    }
    if (!s[r].match(/^[a-z0-9]+$/i)) {
      r--;
      continue;
    }
    if (s[l].toLowerCase() !== s[r].toLowerCase()) {
      return false;
    }

    l++;
    r--;
  }
  return true;
};

File: Valid Palindrome - Leetcode 125/Valid Palindrome - Leetcode 125.java
// Solution 1: O(n)
public class ValidPalindrome {
  public static boolean isPalindrome(String s){
    // Filter out non-alphanumeric
    StringBuilder filtered = new StringBuilder() ;
    for( char c : s.toCharArray())
    {
      if (Character.isLetterOrDigit(c))
      {
        filtered.append(Character.toLowerCase(c));
      }
    }
    // Convert StringBuilder to String
    String str = filtered.toString();

    String reversedStr = filtered.reverse().toString();

    return str.equal(reversedStr)
  }

  public static boolean isPalindromeOptimized(String s){
    int l = 0, r = s.length() - 1;

    while (l<r){
      if(!Character.isLetterOrDigit(s.charAt(l))){
        l++;
        continue;
      }
      if(!Character.isLetterOrDigit(s.charAt(r))){
        r--;
        continue;
      }
      if(Character.toLowerCase(s.charAt(l)) !== Character.toLowerCase(s.charAt(r))){
        return false;
      }
      l++;
      r--;
    }

    return true;
  }
}

public static void main(String[] args){
  System.out.println(isPalindrome("race a car")) // false
}